module/os/operation_siren.py

```diff
@@ -80,12 +81,12 @@ def globe_goto(self, zone, types=('SAFE', 'DANGEROUS'), refresh=False, stop_if_s
         self.globe_update()
         self.globe_focus_to(zone)
         if stop_if_safe:
-            pinned = self.pinned_to_name(self.get_zone_pinned())
-            if pinned == 'SAFE':
+            if self.zone_has_safe():
                 logger.info('Zone is safe, stopped')
                 self.ensure_no_zone_pinned()
                 return False
-        self.zone_type_select(types=types)
+        else:
+            self.zone_type_select(types=types)
```

Comment:
```python
else:
    self.zone_type_select(types=types)
```
Should revert to
```python
self.zone_type_select(types=types)
```
If stop_is_save is True, self.zone_has_safe() is False, and zone has dangerous and obscured zone, it will miss selecting zone and enter the default zone, which is obscured zone.

Reply:
I think this case is already handled? Or I might be missing something? `zone_has_safe` will choose `SELECT_SAFE` if present also returning `True` otherwise `SELECT_DANGEROUS` since that is available in every zone and returns `False`? But it shouldn't need to call `zone_type_select` since it already made the necessary selection. Unless you want it to go through the selection process again? But I would think not?

```
def zone_has_safe(self):
    if self.pinned_to_name(self.get_zone_pinned()) == 'SAFE':
        return True
    else:
        self.ui_click(ZONE_SWITCH, appear_button=self.is_zone_pinned, check_button=self.is_in_zone_select,
                            skip_first_screenshot=True)
        flag = SELECT_SAFE in self.get_zone_select()
        button = SELECT_SAFE if flag else SELECT_DANGEROUS
        self.ui_click(button, check_button=self.is_zone_pinned, offset=self._zone_select_offset,
                        skip_first_screenshot=True)
        return flag
```

The alternative would be to make `zone_has_safe` to making no selection and changing `zone_type_select` to handle when in the zone menu/option or in the pinned state where the zone switch button can be clicked but since it uses `ui_click` primarily that may be difficult.

Reply:
I did realize one hindsight though, in the case there is no `zone_switch` so I've updated to
```
def zone_has_safe(self):
    if self.pinned_to_name(self.get_zone_pinned()) == 'SAFE':
        return True
    elif self.zone_has_switch():
        pass
    else:
        # No zone_switch, already on DANGEROUS
        return False
```

Should be the case if no switch is present at all then its `SELECT_DANGEROUS`, no?
Otherwise in the most minimal case (`SELECT_DANGEROUS` and `SELECT_OBSCURE` the 2nd clause will appropriately `SELECT_DANGEROUS`

Reply:
I think `zone_has_safe` is a method to get infomation from game, and `zone_type_select` is a method to do operations in game. To prevent making thing complecated. A method to get infomation shouldn't do operations in game, vice versa.
`zone_type_select` itself can skip selecting if it's already at targer genre.

Reply:
Understood, the nomenclature should be followed, Thought it save a few clicks and less hassle in doing things twice. I've removed the  `else` and will move onto the server files next. Thanks